### PR TITLE
Refactor drakeCollision C++ library

### PR DESCRIPTION
**Pull request by avalenzu**
* Removes dependence on body indices from drakeCollision, which allows collision
  elements to be created during parsing of a URDF, rather than during RBM
  compilation.
* Removes old tests that used boost's unit test framework
* Provides better separation between RBM and drakeCollision -
  drakeCollision could now be used purely geometric collision detection between collision elements without an RBM or a notion of rigid bodies that contain collision elements.
* Moves the margin/no_margin split into the BulletModel class - RBM no longer
  contains two collision models.

This pull request does not re-implement loading of obj files for mesh objects, and does not remove URDFRigidBodyManipulator.

### CODE REVIEW

**Comment by avalenzu** [comment]
Trying again ...

**Comment by avalenzu** [comment]
Closing to fix tests. Don't know what's up with the windows build servers.

**Comment by avalenzu** [comment]
Fixed the collision filter group passing issues. Not sure what the alignment error from drake004 means:

> c:\\cygwin\   mp\\drake-precompiled-32552\\drake\\systems\\plants\\collision\\Element.h(20): error C2719: 'T_element_to_local': formal parameter with __declspec(align('16')) won't be aligned [C:\\cygwin\   mp\\drake-precompiled-32552\\drake\\pod-build\\systems\\plants\\collision\\drakeCollision.vcxproj]

Also, the CDash page for individual tests (like [this one](http://kobol.csail.mit.edu/cd/testDetails.php?test=1504515&build=11494)) doesn't show the output anymore (at least on my machine). Is that intentional? Or am I looking at the wrong page?

**Comment by avalenzu** [comment]
Closing for now. Will reopen.

**Comment by avalenzu** [comment]
Build and test errors from drake006 should be fixed now.

**Comment by avalenzu** [comment]
@rdeits, have you ever encountered the `FootstepPlan/sanity_check` error that drake001 reported above? Any ideas as to what might cause that?

**Comment by avalenzu** [comment]
@RussTedrake, could you send me login info for drake001 when you get the chance? I was not able to replicate this error on Twan's mac.

File: systems/plants/RigidBody.cpp [hunk 1]
```diff
@@ -28,7 +28,9 @@ RigidBody::RigidBody(void) :
     dSdotVdvi(TWIST_SIZE, 0),
     JdotV(TWIST_SIZE, 1),
     dJdotVdq(TWIST_SIZE, 0),
-    dJdotVdv(TWIST_SIZE, 0)
+    dJdotVdv(TWIST_SIZE, 0),
+    group(DrakeCollision::DEFAULT_GROUP),
```
Comment (RussTedrake) [inline systems/plants/RigidBody.cpp]: should we call these collision_filter_group and collision_filter_mask to be more clear?

File: systems/plants/RigidBody.cpp [hunk 2]
```diff
@@ -146,6 +148,60 @@ bool RigidBody::hasParent() const {
...
+    ids.insert(ids.end(), group_ids_iter->second.begin(), group_ids_iter->second.end());
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool RigidBody::appendCollisionElementIds(vector<DrakeCollision::ElementId>& ids) const
+{
+  ids.reserve(ids.size() + distance(collision_element_ids.begin(), collision_element_ids.end()));
```
Comment (RussTedrake) [inline systems/plants/RigidBody.cpp]: why distance instead of collision.element.ids.size()

File: systems/plants/RigidBody.cpp [hunk 3]
```diff
@@ -146,6 +148,60 @@ bool RigidBody::hasParent() const {
...
+  if (group_ids_iter != collision_element_groups.end()) {
+    ids.reserve(ids.size() + distance(group_ids_iter->second.begin(), group_ids_iter->second.end()));
+    ids.insert(ids.end(), group_ids_iter->second.begin(), group_ids_iter->second.end());
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool RigidBody::appendCollisionElementIds(vector<DrakeCollision::ElementId>& ids) const
```
Comment (RussTedrake) [inline systems/plants/RigidBody.cpp]: Can we document/rename this one?  On quick inspection, I expected it to take ids and append them to the body collision_element_ids, but it does the opposite.

File: systems/plants/RigidBody.m [hunk 4]
```diff
@@ -247,15 +247,13 @@ function dofnum(obj)
 
     function body = makeBelongToCollisionFilterGroup(body,collision_fg_id)
       for id = reshape(collision_fg_id,1,[])
-        body.collision_filter.belongs_to = ...
-          bitor(body.collision_filter.belongs_to,bitshift(1,id-1));
+        body.collision_filter.belongs_to(collision_fg_id) = true;
```
Comment (RussTedrake) [inline systems/plants/RigidBody.m]: out of curiousity, why the change away from the bit logic?  might we be introducing new bugs here?

File: systems/plants/RigidBody.cpp (hunk 1 above)
Comment (avalenzu) [inline systems/plants/RigidBody.cpp]: Agreed

**Comment by avalenzu** [comment]
`appendElementIdsFromThisBody`?

File: systems/plants/collision/CMakeLists.txt [hunk 5]
```diff
@@ -18,5 +18,5 @@ enable_testing()
 add_subdirectory(test)
 
 pods_install_libraries(drakeCollision)
-pods_install_headers(DrakeCollision.h DESTINATION drake/collision)
+pods_install_headers(DrakeCollision.h Geometry.h PointPair.h Element.h Model.h DLLEXPORT_drakeCollision.h DESTINATION drake/collision)
```
Comment (RussTedrake) [inline systems/plants/collision/CMakeLists.txt]: looks like your find and replace might have done extra work here.  (no DLLEXPORT please)

File: systems/plants/collision/DLLEXPORT_drakeCollision.h [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#undef DLLEXPORT_drakeCollision
```
Comment (RussTedrake) [inline systems/plants/collision/DLLEXPORT_drakeCollision.h]: change this filename, too.

File: systems/plants/collision/DrakeCollision.h [hunk 7]
```diff
@@ -9,171 +9,26 @@
...
-#if defined(WIN32) || defined(WIN64)
-  #if defined(drakeCollision_EXPORTS)
-    #define DLLEXPORT __declspec( dllexport )
-  #else
-    #define DLLEXPORT __declspec( dllimport )
-  #endif
-#else
-    #define DLLEXPORT
-#endif
+#include "DLLEXPORT_drakeCollision.h"
```
Comment (RussTedrake) [inline systems/plants/collision/DrakeCollision.h]: same.

File: systems/plants/collision/DrakeCollision.h [hunk 8]
```diff
@@ -9,171 +9,26 @@
...
-
-  DLLEXPORT std::shared_ptr<Model> newModel(ModelType model_type);
-
-
-  
-  typedef std::bitset<16> bitmask;
   // Constants
-  extern const DLLEXPORT bitmask ALL_MASK;
-  extern const DLLEXPORT bitmask NONE_MASK;
-  extern const DLLEXPORT bitmask DEFAULT_GROUP;
```
Comment (RussTedrake) [inline systems/plants/collision/DrakeCollision.h]: should these be DLLEXPORT_drakeCollision?

File: systems/plants/collision/DrakeCollision.h [hunk 9]
```diff
@@ -9,171 +9,26 @@
...
-    //
-    virtual bool collisionRaycast(const Eigen::Matrix3Xd &origin, const Eigen::Matrix3Xd &ray_endpoint, Eigen::VectorXd &distances) { return false; };
+  typedef std::bitset<MAX_NUM_COLLISION_FILTER_GROUPS> bitmask;
 
-    protected:
-      virtual const std::vector<int> bodyIndices() const;
-      virtual const std::set<std::string> elementGroupNames() const;
-  };
-
-  DLLEXPORT std::shared_ptr<Model> newModel();
```
Comment (RussTedrake) [inline systems/plants/collision/DrakeCollision.h]: wrong DLLEXPORT?

File: systems/plants/RigidBody.cpp (hunk 2 above)
Comment (avalenzu) [inline systems/plants/RigidBody.cpp]: Yes, that'd be better

File: systems/plants/RigidBody.cpp (hunk 3 above)
Comment (avalenzu) [inline systems/plants/RigidBody.cpp]: `addCollisionElementIdsFromThisBody`?

**Comment by RussTedrake** [comment]
otherwise looks good.

File: systems/plants/collision/DLLEXPORT_drakeCollision.h (hunk 6 above)
Comment (avalenzu) [inline systems/plants/collision/DLLEXPORT_drakeCollision.h]: Why? This file exists to contain the DLLEXPORT_drakeCollision macro so that I don't have it copied multiple places.

File: systems/plants/collision/DrakeCollision.h (hunk 9 above)
Comment (avalenzu) [inline systems/plants/collision/DrakeCollision.h]: Ouch. Yes. Will fix.

File: systems/plants/RigidBody.m (hunk 4 above)
Comment (avalenzu) [inline systems/plants/RigidBody.m]: I was running into bugs caused by converting MATLAB integer types into `std::bitset` objects. The logic is the same, really, it's just that now the MATLAB side properties are logical vectors. In `constructModelmex` I loop over those vectors and set the appropriate bits in the `std::bitset`.

File: systems/plants/collision/DrakeCollision.h (hunk 9 above)
Comment (avalenzu) [inline systems/plants/collision/DrakeCollision.h]: On second look, this is fine - those lines with the wrong DLLEXPORT are being removed by the diff, not added.

File: systems/plants/collision/DLLEXPORT_drakeCollision.h (hunk 6 above)
Comment (avalenzu) [inline systems/plants/collision/DLLEXPORT_drakeCollision.h]: Would it be better to just copy the macro to the relevant headers?

File: systems/plants/RigidBody.cpp (hunk 3 above)
Comment (avalenzu) [inline systems/plants/RigidBody.cpp]: Or better, `appendCollisionElementIdsFromThisBody`

**Comment by avalenzu** [comment]
OK, we've got positive test results from build servers for each platform.
The error in runDRCDoorTask did not repeat on drake001.

**Comment by RussTedrake** [comment]
Sorry @avalenzu , but could you merge w/ master for me?  (@psiorx's pull request has been merged upstream 5-6 times already, so I merged it first).

**Comment by avalenzu** [comment]
Will do

**Comment by avalenzu** [comment]
Ok, master is merged back in. If unit test pass, I think this is good to go.

**Comment by avalenzu** [comment]
Closing this - its changes are intertwined with the ones I'm making to get RobotLocomotion/master and drc working together. I think it'll be cleaner to bring them in together.
